### PR TITLE
Connector-less usage

### PR DESCRIPTION
**Pull request by JasonStoltz**
## Description

This PR implements Connector-less usage, which enables users to use Search UI by providing individual event handlers, rather than using a Connector.

Much of the design is captured here: https://github.com/elastic/search-ui/issues/110.

Here are the ways in which you can now configure Search UI:

### With a Connector

```jsx
<SearchProvider
  config={{
    apiConnector: connector
  }}
/>
```

### Without a Connector (Connector-less)

```jsx
<SearchProvider
  config={{
    onSearch: async state => {
      const queryForOtherService = transformSearchUIStateToQuery(state);
      const otherServiceResponse = await callSomeOtherService(
        queryForOtherService
      );
      return transformOtherServiceResponseToSearchUIState(otherServiceResponse);
    }
    onAutocomplete() => { /* handler */ },
    onResultClick() => { /* handler */ },
    onAutocompleteResultClick() => { /* handler */ },
  }}
/>
```

### With a Connector, but overriding an individual Handler

```jsx
<SearchProvider
  config={{
    apiConnector: connector,
    onSearch: async (state, queryConfig) => {
      const queryForOtherService = transformSearchUIStateToQuery(
        state,
        queryConfig
      );
      const otherServiceResponse = await callSomeOtherService(
        queryForOtherService
      );
      return transformOtherServiceResponseToSearchUIState(otherServiceResponse);
    }
  }}
/>
```

### With a Connector, but enhancing an individual Handler by using it as middleware

```jsx
<SearchProvider
  config={{
    apiConnector: connector,
    onSearch: (state, queryConfig, next) => {
      const updatedState = someStateTransformation(state);
      return next(updatedState, queryConfig);
    }
  }}
/>
```

## List of changes

- Created an "Events" object in SearchDriver to decouple handler logic from the driver itself.
- Renamed handler events in the connectors.
- Added handler hooks directly to `SearchDriver` as configuration, which enables connector-less usage, as well as overriding and enhancing Connector usage

### CODE REVIEW

File: ADVANCED.md [hunk 1]
```diff
@@ -811,16 +811,26 @@ return (
...
-| `searchQuery`       | [Query Config](#query-config)                                           | optional  |         | Configuration options for the main search query.                                                                                                                                                                                                         |
-| `trackUrlState`     | boolean                                                                 | optional  |         | By default, [Request State](#headless-core#1-request-state) will be synced with the browser url. To turn this off, pass `false`.                                                                                                                         |
+apiConnector: connector,
+// Example of overriding a connector handler
+onResultClick: () => console.log("hi"),
+// Example of handler as middleware
+onSearch: (state, queryConfig, next) => next(state, queryConfig)
```
Comment (omegahm) [inline ADVANCED.md]: What are these examples? They don't seem to belong here?

**Comment by omegahm** [comment]
Looks good. Nice commit structure. :happy:

File: ADVANCED.md (hunk 1 above)
Comment (JasonStoltz) [reply ADVANCED.md]: 🤦‍♂ Good catch, those are NOT supposed to be there. Will update.

**Comment by JasonStoltz** [comment]
@omegahm In what sense did you like the commit structure? Did it makes it easier to review? I typically just end up Squash + Merging these, just FYI.

**Comment by omegahm** [comment]
It was nice that the commits were small and all the changes in them where coherent, even though you did end up renaming the same property three times. 😅

**Pull request merged**